File: frontend/lib/models/appointment.dart
import 'package:json_annotation/json_annotation.dart';
import 'client.dart';

part 'appointment.g.dart';

@JsonSerializable()
class Appointment {
  final String id;
  final String userId;
  final String clientId;
  final Client? client;
  final String title;
  final String? description;
  final DateTime startDateTime;
  final DateTime endDateTime;
  final String? location;
  final String appointmentType;
  final String status;
  final String priority;
  final int reminderMinutes;
  final bool isRecurring;
  final RecurringPattern? recurringPattern;
  final List<Attendee>? attendees;
  final List<AppointmentNote>? notes;
  final List<Document>? documents;
  final double billableHours;
  final double? hourlyRate;
  final double totalAmount;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  Appointment({
    required this.id,
    required this.userId,
    required this.clientId,
    this.client,
    required this.title,
    this.description,
    required this.startDateTime,
    required this.endDateTime,
    this.location,
    required this.appointmentType,
    required this.status,
    required this.priority,
    required this.reminderMinutes,
    required this.isRecurring,
    this.recurringPattern,
    this.attendees,
    this.notes,
    this.documents,
    required this.billableHours,
    this.hourlyRate,
    required this.totalAmount,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Appointment.fromJson(Map<String, dynamic> json) => _$AppointmentFromJson(json);
  Map<String, dynamic> toJson() => _$AppointmentToJson(this);

  Duration get duration => endDateTime.difference(startDateTime);
  
  int get durationMinutes => duration.inMinutes;

  String get dateRange {
    final startDate = startDateTime.toLocal();
    final endDate = endDateTime.toLocal();
    
    if (startDate.day == endDate.day && 
        startDate.month == endDate.month && 
        startDate.year == endDate.year) {
      return '${startDate.day}/${startDate.month}/${startDate.year}';
    }
    
    return '${startDate.day}/${startDate.month}/${startDate.year} - ${endDate.day}/${endDate.month}/${endDate.year}';
  }

  Appointment copyWith({
    String? id,
    String? userId,
    String? clientId,
    Client? client,
    String? title,
    String? description,
    DateTime? startDateTime,
    DateTime? endDateTime,
    String? location,
    String? appointmentType,
    String? status,
    String? priority,
    int? reminderMinutes,
    bool? isRecurring,
    RecurringPattern? recurringPattern,
    List<Attendee>? attendees,
    List<AppointmentNote>? notes,
    List<Document>? documents,
    double? billableHours,
    double? hourlyRate,
    double? totalAmount,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Appointment(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      clientId: clientId ?? this.clientId,
      client: client ?? this.client,
      title: title ?? this.title,
      description: description ?? this.description,
      startDateTime: startDateTime ?? this.startDateTime,
      endDateTime: endDateTime ?? this.endDateTime,
      location: location ?? this.location,
      appointmentType: appointmentType ?? this.appointmentType,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      reminderMinutes: reminderMinutes ?? this.reminderMinutes,
      isRecurring: isRecurring ?? this.isRecurring,
      recurringPattern: recurringPattern ?? this.recurringPattern,
      attendees: attendees ?? this.attendees,
      notes: notes ?? this.notes,
      documents: documents ?? this.documents,
      billableHours: billableHours ?? this.billableHours,
      hourlyRate: hourlyRate ?? this.hourlyRate,
      totalAmount: totalAmount ?? this.totalAmount,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

@JsonSerializable()
class RecurringPattern {
  final String frequency;
  final int interval;
  final DateTime? endDate;
  final List<int>? daysOfWeek;

  RecurringPattern({
    required this.frequency,
    required this.interval,
    this.endDate,
    this.daysOfWeek,
  });

  factory RecurringPattern.fromJson(Map<String, dynamic> json) => _$RecurringPatternFromJson(json);
  Map<String, dynamic> toJson() => _$RecurringPatternToJson(this);
}

@JsonSerializable()
class Attendee {
  final String name;
  final String? email;
  final String? phone;
  final String role;

  Attendee({
    required this.name,
    this.email,
    this.phone,
    required this.role,
  });

  factory Attendee.fromJson(Map<String, dynamic> json) => _$AttendeeFromJson(json);
  Map<String, dynamic> toJson() => _$AttendeeToJson(this);
}

@JsonSerializable()
class AppointmentNote {
  final String content;
  final DateTime createdAt;

  AppointmentNote({
    required this.content,
    required this.createdAt,
  });

  factory AppointmentNote.fromJson(Map<String, dynamic> json) => _$AppointmentNoteFromJson(json);
  Map<String, dynamic> toJson() => _$AppointmentNoteToJson(this);
}

File: backend/models/Appointment.js
const mongoose = require('mongoose');

const appointmentSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  clientId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Client',
    required: true
  },
  title: {
    type: String,
    required: [true, 'Appointment title is required'],
    trim: true,
    maxlength: [200, 'Title cannot exceed 200 characters']
  },
  description: {
    type: String,
    maxlength: [1000, 'Description cannot exceed 1000 characters']
  },
  startDateTime: {
    type: Date,
    required: [true, 'Start date and time is required']
  },
  endDateTime: {
    type: Date,
    required: [true, 'End date and time is required']
  },
  location: {
    type: String,
    trim: true,
    maxlength: [200, 'Location cannot exceed 200 characters']
  },
  appointmentType: {
    type: String,
    enum: ['Consultation', 'Court Hearing', 'Client Meeting', 'Document Review', 'Mediation', 'Deposition', 'Other'],
    default: 'Consultation'
  },
  status: {
    type: String,
    enum: ['Scheduled', 'Confirmed', 'In Progress', 'Completed', 'Cancelled', 'Rescheduled'],
    default: 'Scheduled'
  },
  priority: {
    type: String,
    enum: ['Low', 'Medium', 'High', 'Urgent'],
    default: 'Medium'
  },
  reminderMinutes: {
    type: Number,
    default: 30,
    min: [0, 'Reminder minutes cannot be negative']
  },
  isRecurring: {
    type: Boolean,
    default: false
  },
  recurringPattern: {
    frequency: {
      type: String,
      enum: ['Daily', 'Weekly', 'Monthly', 'Yearly']
    },
    interval: {
      type: Number,
      min: 1
    },
    endDate: Date,
    daysOfWeek: [Number] // 0-6, Sunday to Saturday
  },
  attendees: [{
    name: String,
    email: String,
    phone: String,
    role: {
      type: String,
      enum: ['Client', 'Lawyer', 'Witness', 'Expert', 'Other'],
      default: 'Client'
    }
  }],
  notes: [{
    content: String,
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  documents: [{
    name: String,
    path: String,
    uploadedAt: {
      type: Date,
      default: Date.now
    }
  }],
  billableHours: {
    type: Number,
    min: [0, 'Billable hours cannot be negative'],
    default: 0
  },
  hourlyRate: {
    type: Number,
    min: [0, 'Hourly rate cannot be negative']
  },
  totalAmount: {
    type: Number,
    min: [0, 'Total amount cannot be negative'],
    default: 0
  },
  isActive: {
    type: Boolean,
    default: true
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Create compound indexes for efficient querying
appointmentSchema.index({ userId: 1, startDateTime: 1 });
appointmentSchema.index({ userId: 1, clientId: 1 });
appointmentSchema.index({ userId: 1, status: 1 });
appointmentSchema.index({ startDateTime: 1, endDateTime: 1 });

// Validation: End time must be after start time
appointmentSchema.pre('save', function(next) {
  if (this.endDateTime <= this.startDateTime) {
    return next(new Error('End time must be after start time'));
  }
  
  // Calculate total amount if billable hours and hourly rate are provided
  if (this.billableHours && this.hourlyRate) {
    this.totalAmount = this.billableHours * this.hourlyRate;
  }
  
  this.updatedAt = Date.now();
  next();
});

// Virtual for duration in minutes
appointmentSchema.virtual('durationMinutes').get(function() {
  return Math.round((this.endDateTime - this.startDateTime) / (1000 * 60));
});

// Virtual for formatted date range
appointmentSchema.virtual('dateRange').get(function() {
  const start = this.startDateTime.toLocaleDateString();
  const end = this.endDateTime.toLocaleDateString();
  return start === end ? start : `${start} - ${end}`;
});

module.exports = mongoose.model('Appointment', appointmentSchema);
